File: packages/database/prisma/seed.ts
import {
  AdminRole,
  PaymentProvider,
  PrismaClient,
  QualityMode,
  SubscriptionStatus,
} from "@prisma/client";
import argon2 from "argon2";

const prisma = new PrismaClient();

const plans = [
  {
    key: "trial",
    name: "Free Trial",
    description: "A consent-first 30 second AI face call trial.",
    monthlyCredits: 1_500,
    maxFaceProfiles: 1,
    maxImagesPerProfile: 1,
    maxParticipants: 2,
    maxCallMinutes: 1,
    maxGroupCalls: 0,
    watermarkRequired: true,
    creditTopupsAllowed: false,
    creditResetDays: 7,
    allowedQualities: [QualityMode.LOW, QualityMode.STANDARD],
    groupCalls: false,
    voiceEffects: false,
    cloudGpu: false,
    priceMonthlyMinor: 0,
    sortOrder: 0,
  },
  {
    key: "basic",
    name: "Basic",
    description: "Standard quality for regular one-to-one calls.",
    monthlyCredits: 30_000,
    maxFaceProfiles: 3,
    maxImagesPerProfile: 2,
    maxParticipants: 2,
    maxCallMinutes: 60,
    maxGroupCalls: 0,
    watermarkRequired: true,
    creditTopupsAllowed: true,
    creditResetDays: 30,
    allowedQualities: [QualityMode.LOW, QualityMode.STANDARD],
    groupCalls: false,
    voiceEffects: true,
    cloudGpu: false,
    priceMonthlyMinor: 1_900,
    sortOrder: 1,
  },
  {
    key: "pro",
    name: "Pro",
    description: "HD, group calls, and priority browser processing.",
    monthlyCredits: 120_000,
    maxFaceProfiles: 10,
    maxImagesPerProfile: 5,
    maxParticipants: 12,
    maxCallMinutes: 240,
    maxGroupCalls: 100,
    watermarkRequired: true,
    creditTopupsAllowed: true,
    creditResetDays: 30,
    allowedQualities: [QualityMode.LOW, QualityMode.STANDARD, QualityMode.HD],
    groupCalls: true,
    voiceEffects: true,
    cloudGpu: false,
    priceMonthlyMinor: 4_900,
    sortOrder: 2,
  },
  {
    key: "business",
    name: "Business",
    description: "Teams, branding, analytics, and configurable limits.",
    monthlyCredits: 500_000,
    maxFaceProfiles: 50,
    maxImagesPerProfile: 10,
    maxParticipants: 50,
    maxCallMinutes: 720,
    maxGroupCalls: 1000,
    watermarkRequired: false,
    creditTopupsAllowed: true,
    creditResetDays: 30,
    allowedQualities: [QualityMode.LOW, QualityMode.STANDARD, QualityMode.HD],
    groupCalls: true,
    voiceEffects: true,
    cloudGpu: true,
    priceMonthlyMinor: 14_900,
    sortOrder: 3,
  },
] as const;

async function main() {
  for (const plan of plans) {
    await prisma.plan.upsert({
      where: { key: plan.key },
      update: plan,
      create: plan,
    });
  }

  await prisma.appSetting.upsert({
    where: {
      namespace_key: { namespace: "branding", key: "public" },
    },
    update: {},
    create: {
      namespace: "branding",
      key: "public",
      publicValue: {
        appName: "TrueFace Calls",
        supportEmail: "support@example.com",
        primaryColor: "#4f46e5",
      },
    },
  });

  const defaultSettings = [
    {
      namespace: "billing",
      key: "credit-packs",
      publicValue: {
        packs: [
          {
            key: "starter",
            name: "Starter credits",
            creditsMilli: 25_000,
            amountMinor: 500,
            currency: "USD",
          },
          {
            key: "creator",
            name: "Creator credits",
            creditsMilli: 120_000,
            amountMinor: 2_000,
            currency: "USD",
          },
        ],
      },
    },
    {
      namespace: "billing",
      key: "usage-rates",
      publicValue: {
        baseCallMilliPerMinute: 250,
        aiFaceMilliPerMinute: 1_000,
        voiceEffectMultiplier: 1.25,
        hdMultiplier: 1.5,
        cloudGpuMultiplier: 2.5,
      },
    },
    {
      namespace: "features",
      key: "flags",
      publicValue: {
        guestJoin: true,
        screenShare: true,
        voiceEffects: false,
        cloudGpu: false,
        manualBankTransfer: false,
      },
    },
    {
      namespace: "safety",
      key: "policy",
      publicValue: {
        abuseReportingEnabled: true,
        faceModerationRequired: true,
        consentTermsVersion: "2026-06",
        deleteRejectedFaceMedia: true,
      },
    },
  ];

  for (const setting of defaultSettings) {
    await prisma.appSetting.upsert({
      where: {
        namespace_key: {
          namespace: setting.namespace,
          key: setting.key,
        },
      },
      update: {},
      create: setting,
    });
  }

  const adminEmail = process.env.BOOTSTRAP_ADMIN_EMAIL;
  const adminPassword = process.env.BOOTSTRAP_ADMIN_PASSWORD;
  const previewEmail = process.env.BOOTSTRAP_PREVIEW_EMAIL;
  const previewPassword = process.env.BOOTSTRAP_PREVIEW_PASSWORD;
  const trial = await prisma.plan.findUniqueOrThrow({
    where: { key: "trial" },
  });

  if (adminEmail && adminPassword) {
    const user = await ensureBootstrapUser(
      adminEmail,
      adminPassword,
      "Platform Admin",
      trial.id,
      trial.monthlyCredits,
    );

    await prisma.adminUser.upsert({
      where: { userId: user.id },
      update: { active: true },
      create: {
        userId: user.id,
        role: AdminRole.SUPER_ADMIN,
        permissions: ["*"],
      },
    });
  }

  if (previewEmail && previewPassword) {
    await ensureBootstrapUser(
      previewEmail,
      previewPassword,
      "Preview Viewer",
      trial.id,
      trial.monthlyCredits,
    );
  }
}

async function ensureBootstrapUser(
  email: string,
  password: string,
  displayName: string,
  planId: string,
  trialCredits: number,
) {
  const normalizedEmail = email.toLowerCase();
  const user = await prisma.user.upsert({
    where: { email: normalizedEmail },
    update: {
      displayName,
      emailVerifiedAt: new Date(),
    },
    create: {
      email: normalizedEmail,
      displayName,
      passwordHash: await argon2.hash(password, {
        type: argon2.argon2id,
      }),
      emailVerifiedAt: new Date(),
    },
  });

  const wallet = await prisma.creditWallet.upsert({
    where: { userId: user.id },
    update: {},
    create: {
      userId: user.id,
      availableMilliCredits: trialCredits,
      includedMilliCredits: trialCredits,
      includedResetAt: new Date(Date.now() + 7 * 24 * 60 * 60 * 1000),
    },
  });

  const existingSubscription = await prisma.subscription.findFirst({
    where: { userId: user.id },
  });
  if (!existingSubscription) {
    await prisma.subscription.create({
      data: {
        userId: user.id,
        planId,
        provider: PaymentProvider.MANUAL,
        status: SubscriptionStatus.TRIALING,
        currentPeriodStart: new Date(),
        currentPeriodEnd: new Date(Date.now() + 7 * 24 * 60 * 60 * 1000),
      },
    });
  }

  await prisma.creditTransaction.upsert({
    where: { idempotencyKey: `bootstrap:${user.id}` },
    update: {},
    create: {
      walletId: wallet.id,
      type: "TRIAL_GRANT",
      amountMilli: trialCredits,
      balanceAfterMilli: wallet.availableMilliCredits,
      source: "bootstrap",
      idempotencyKey: `bootstrap:${user.id}`,
    },
  });

  return user;
}

main()
  .finally(async () => {
    await prisma.$disconnect();
  })
  .catch((error: unknown) => {
    console.error(error);
    process.exitCode = 1;
  });

File: backend/app/seed.py
from __future__ import annotations

import uuid
from datetime import timedelta

from argon2 import PasswordHasher
from pymongo.database import Database

from .serializers import utc_now

PASSWORDS = PasswordHasher()

DEFAULT_PLANS = [
    {
        "key": "trial",
        "name": "Free Trial",
        "description": "Short consent-first product evaluation.",
        "monthlyCredits": 1500,
        "maxFaceProfiles": 1,
        "maxImagesPerProfile": 2,
        "maxParticipants": 2,
        "maxCallMinutes": 1,
        "maxGroupCalls": 0,
        "allowedQualities": ["LOW", "STANDARD"],
        "groupCalls": False,
        "watermarkRequired": True,
        "creditTopupsAllowed": False,
        "creditResetDays": 30,
        "voiceEffects": False,
        "cloudGpu": False,
        "priceMonthlyMinor": 0,
        "currency": "NGN",
        "sortOrder": 0,
        "enabled": True,
    },
    {
        "key": "basic",
        "name": "Basic",
        "description": "Standard individual AI calling.",
        "monthlyCredits": 60000,
        "maxFaceProfiles": 3,
        "maxImagesPerProfile": 5,
        "maxParticipants": 2,
        "maxCallMinutes": 60,
        "maxGroupCalls": 0,
        "allowedQualities": ["LOW", "STANDARD"],
        "groupCalls": False,
        "watermarkRequired": True,
        "creditTopupsAllowed": True,
        "creditResetDays": 30,
        "voiceEffects": False,
        "cloudGpu": False,
        "priceMonthlyMinor": 990_000,
        "currency": "NGN",
        "sortOrder": 10,
        "enabled": True,
    },
    {
        "key": "pro",
        "name": "Pro",
        "description": "HD and group calls for creators and professionals.",
        "monthlyCredits": 240000,
        "maxFaceProfiles": 10,
        "maxImagesPerProfile": 8,
        "maxParticipants": 8,
        "maxCallMinutes": 240,
        "maxGroupCalls": 100,
        "allowedQualities": ["LOW", "STANDARD", "HD"],
        "groupCalls": True,
        "watermarkRequired": True,
        "creditTopupsAllowed": True,
        "creditResetDays": 30,
        "voiceEffects": True,
        "cloudGpu": True,
        "priceMonthlyMinor": 2_490_000,
        "currency": "NGN",
        "sortOrder": 20,
        "enabled": True,
    },
    {
        "key": "business",
        "name": "Business",
        "description": "Team controls, analytics, and priority processing.",
        "monthlyCredits": 1000000,
        "maxFaceProfiles": 50,
        "maxImagesPerProfile": 10,
        "maxParticipants": 25,
        "maxCallMinutes": 480,
        "maxGroupCalls": 1000,
        "allowedQualities": ["LOW", "STANDARD", "HD"],
        "groupCalls": True,
        "watermarkRequired": True,
        "creditTopupsAllowed": True,
        "creditResetDays": 30,
        "voiceEffects": True,
        "cloudGpu": True,
        "priceMonthlyMinor": 7_490_000,
        "currency": "NGN",
        "sortOrder": 30,
        "enabled": True,
    },
]

DEFAULT_CREDIT_PACKS = [
    {
        "key": "starter",
        "name": "Starter credits",
        "creditsMilli": 50_000,
        "priceMinor": 250_000,
        "currency": "NGN",
    },
    {
        "key": "growth",
        "name": "Growth credits",
        "creditsMilli": 250_000,
        "priceMinor": 1_000_000,
        "currency": "NGN",
    },
    {
        "key": "scale",
        "name": "Scale credits",
        "creditsMilli": 1_000_000,
        "priceMinor": 3_500_000,
        "currency": "NGN",
    },
]


def ensure_indexes(db: Database) -> None:
    db.users.create_index("email", unique=True)
    db.sessions.create_index("refreshTokenHash", unique=True)
    db.auth_tokens.create_index("tokenHash", unique=True)
    db.plans.create_index("key", unique=True)
    db.subscriptions.create_index("id", unique=True)
    db.call_rooms.create_index("slug", unique=True)
    db.credit_wallets.create_index("userId", unique=True)
    db.credit_transactions.create_index("idempotencyKey", unique=True)
    db.usage_minutes.create_index("meteringWindow", unique=True)
    db.payments.create_index("idempotencyKey", unique=True)
    db.payments.create_index("paymentReference", unique=True, sparse=True)
    db.webhook_events.create_index(
        [("provider", 1), ("externalId", 1)], unique=True, sparse=True
    )
    db.app_settings.create_index([("namespace", 1), ("key", 1)], unique=True)
    db.provider_health.create_index("provider", unique=True)
    db.blocked_users.create_index([("blockerId", 1), ("blockedId", 1)], unique=True)
    db.notification_reads.create_index(
        [("notificationId", 1), ("userId", 1)], unique=True
    )
    db.rate_limits.create_index("expiresAt", expireAfterSeconds=0)


def seed_database(
    db: Database,
    admin_email: str | None = None,
    admin_password: str | None = None,
) -> None:
    ensure_indexes(db)
    now = utc_now()
    for plan in DEFAULT_PLANS:
        existing = db.plans.find_one({"key": plan["key"]})
        if not existing:
            db.plans.insert_one(
                {
                    **plan,
                    "id": str(uuid.uuid4()),
                    "createdAt": now,
                    "updatedAt": now,
                }
            )
        elif existing.get("currency") != "NGN":
            db.plans.update_one(
                {"id": existing["id"]},
                {
                    "$set": {
                        "currency": "NGN",
                        "priceMonthlyMinor": plan["priceMonthlyMinor"],
                        "updatedAt": now,
                    }
                },
            )
    credit_setting = db.app_settings.find_one(
        {"namespace": "billing", "key": "credit-packs"}
    )
    if not credit_setting:
        db.app_settings.insert_one(
            {
                "id": str(uuid.uuid4()),
                "namespace": "billing",
                "key": "credit-packs",
                "publicValue": {"packs": DEFAULT_CREDIT_PACKS},
                "version": 1,
                "createdAt": now,
                "updatedAt": now,
            }
        )
    else:
        packs = credit_setting.get("publicValue", {}).get("packs", [])
        defaults = {pack["key"]: pack for pack in DEFAULT_CREDIT_PACKS}
        migrated = []
        changed = False
        for pack in packs:
            if pack.get("currency") == "NGN":
                migrated.append(pack)
                continue
            default = defaults.get(pack.get("key"))
            migrated.append(default or {**pack, "currency": "NGN"})
            changed = True
        if changed:
            db.app_settings.update_one(
                {"_id": credit_setting["_id"]},
                {
                    "$set": {
                        "publicValue.packs": migrated,
                        "updatedAt": now,
                    }
                },
            )
    if not admin_email or not admin_password:
        return
    email = admin_email.strip().lower()
    user = db.users.find_one({"email": email})
    if not user:
        user_id = str(uuid.uuid4())
        user = {
            "id": user_id,
            "email": email,
            "passwordHash": PASSWORDS.hash(admin_password),
            "displayName": "TrueFace Administrator",
            "status": "ACTIVE",
            "emailVerifiedAt": now,
            "locale": "en",
            "timezone": "UTC",
            "roomCreationDisabled": False,
            "createdAt": now,
            "updatedAt": now,
        }
        db.users.insert_one(user)
        trial = db.plans.find_one({"key": "trial"})
        db.subscriptions.insert_one(
            {
                "id": str(uuid.uuid4()),
                "userId": user_id,
                "planId": trial["id"],
                "provider": "MANUAL",
                "status": "TRIALING",
                "currentPeriodStart": now,
                "currentPeriodEnd": now + timedelta(days=5),
                "createdAt": now,
                "updatedAt": now,
            }
        )
        db.credit_wallets.insert_one(
            {
                "id": str(uuid.uuid4()),
                "userId": user_id,
                "availableMilliCredits": trial["monthlyCredits"],
                "reservedMilliCredits": 0,
                "includedMilliCredits": trial["monthlyCredits"],
                "purchasedMilliCredits": 0,
                "reservedIncludedMilli": 0,
                "reservedPurchasedMilli": 0,
                "includedResetAt": now + timedelta(days=30),
                "lifetimePurchasedMilli": 0,
                "lifetimeConsumedMilli": 0,
                "version": 0,
                "createdAt": now,
                "updatedAt": now,
            }
        )
    db.admin_users.update_one(
        {"userId": user["id"]},
        {
            "$setOnInsert": {
                "id": str(uuid.uuid4()),
                "userId": user["id"],
                "role": "SUPER_ADMIN",
                "permissions": ["*"],
                "mfaRequired": True,
                "mfaEnabled": False,
                "active": True,
                "createdAt": now,
                "updatedAt": now,
            }
        },
        upsert=True,
    )
